File: tasks/circuit-breaker/python/src/parse.py
"""Read the configuration and classify call outcomes."""

TOP_LEVEL = {"config", "calls"}


def parse_document(value):
    """Accept the parsed input as a run description.

    The current implementation checks only the top-level key set and trusts
    every value below it.
    """
    if not isinstance(value, dict) or set(value) != TOP_LEVEL:
        raise ValueError("malformed document")
    return value


def parse_config(value):
    return {
        "threshold": value["threshold"],
        "cooldownMs": value["cooldownMs"],
        "halfOpenLimit": value["halfOpenLimit"],
        "failureStatuses": value["failureStatuses"],
    }


def classify(outcome, failure_statuses):
    """Every outcome that is not a success counts as a failure."""
    return "success" if outcome["kind"] == "ok" else "failure"

File: tasks/circuit-breaker/go/src/parse.go
package main

// callOutcome is the result a call reported.
type callOutcome struct {
	Kind   string `json:"kind"`
	Status int    `json:"status"`
}

type call struct {
	At      int         `json:"at"`
	Target  string      `json:"target"`
	Outcome callOutcome `json:"outcome"`
}

type config struct {
	Threshold       int   `json:"threshold"`
	CooldownMs      int   `json:"cooldownMs"`
	HalfOpenLimit   int   `json:"halfOpenLimit"`
	FailureStatuses []int `json:"failureStatuses"`
}

type document struct {
	Config config `json:"config"`
	Calls  []call `json:"calls"`
}

// classify treats every outcome that is not a success as a failure.
func classify(outcome callOutcome, failureStatuses []int) string {
	if outcome.Kind == "ok" {
		return "success"
	}
	return "failure"
}
